server/src/main/java/org/opensearch/index/shard/SearchOperationListener.java

```diff
@@ -71,6 +71,30 @@ default void onFailedQueryPhase(SearchContext searchContext) {}
      */
     default void onQueryPhase(SearchContext searchContext, long tookInNanos) {}
 
+    /**
+     * Executed before the slice execution in
```

Comment:
Will update the comment. 

What sort of information about the segments that a given slice has worked on do you think would be useful?

Reply:
I guess there's metadata related to each LeafReaderContext that may be useful so I think we might as well add it for now then.

Reply:
Added to the latest revision, @sohami let me know what you think!

Reply:
@sohami @jed326 sorry to jump in later, the listener is supposed to notify about search stages, not provide full fledged hooks (you can see that this applicable to every single method in listener). If we need more control over the notification and its data - let's have a dedicated listener.

Reply:
Thanks @reta -- I understand what you're saying (and I think I agree with it too), but at the same time I also don't really see the difference between the stage notification vs hooks since the other existing listener methods are passing the entire `searchContext` as well and doing processing on that context so that doesn't seem any different from taking the list of leaves and doing any processing with that.

That being said, I'm fine with removing the leaves reference as I don't have a specific use case for it today and I was including it more as a "just in case I need it".

Reply:
> That being said, I'm fine with removing the leaves reference as I don't have a specific use case for it today and I was including it more as a "just in case I need it".

Thanks @jed326 , the difference is that we bring Apache Lucene as a direct dependency here, it feels like something is expected to be done with leaves but what exactly? Exposing the internals just because we could should not be the driver I think.

Reply:
That's fair, I've removed the `leaves` parameter in the latest revision. @sohami let me know what you think

Reply:
> Thanks @jed326 , the difference is that we bring Apache Lucene as a direct dependency here, it feels like something is expected to be done with leaves but what exactly? Exposing the internals just because we could should not be the driver I think.

I had similar concern but wanted to hear other thoughts. Other way I was thinking was to expose some information via another object instead of exposing lucene dependency. Few information which I think could be useful are: i) number of leaves per slice, ii) number of docs per leaf per slice, iii) size of leaves per slice. But I guess for now I am fine to keep it simple and we can add it later as needed.